aptos-move/aptos-vm/src/aptos_vm.rs

```diff
@@ -1357,14 +1356,31 @@ impl AptosVM {
                 //       result in shallow-loading of the modules and therefore subtle changes in
                 //       the error semantics.
                 if self.gas_feature_version >= 15 {
-                    // Charge old versions of the modules, in case of upgrades.
-                    session.check_dependencies_and_charge_gas_non_recursive_optional(
-                        gas_meter,
-                        traversal_context,
-                        modules
-                            .iter()
-                            .map(|module| (module.self_addr(), module.self_name())),
-                    )?;
+                    // Charge old versions of existing modules, in case of upgrades.
+                    for module in modules.iter() {
+                        let addr = module.self_addr();
+                        let name = module.self_name();
+                        let state_key = StateKey::module(addr, name);
+
+                        if let Some(size) = resolver
+                            .as_executor_view()
+                            .get_module_state_value_size(&state_key)
+                            .map_err(|e| e.finish(Location::Undefined))?
+                        {
+                            if !addr.is_special()
```

Comment:
nit: the previous way (continue when this condition is negative) was more readable (less indent)

Reply:
done